File: leetcode/283. Move Zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int zeros = 0;
        int size = nums.size();
        int index = 0;
        while(index < size){
            if(nums[index] == 0){
                nums.erase(nums.begin()+index);
                zeros++;
            }else{
                index++;
            }
            size = nums.size();
        }
        while(zeros>0){
            nums.push_back(0);
            zeros--;
        }
    }
};

File: online-training-platform/leetcode/283. Move Zeroes.kt
class Solution {
    fun moveZeroes(nums: IntArray): Unit {
        
        var zeroIdx = nums.indexOfFirst{
            it == 0
        }
        var nonZeroIdx = nums.indexOfFirst {
            it != 0
        }

        if(zeroIdx == -1 || nonZeroIdx == -1) {
            return 
        }


        while(true){
            if( zeroIdx < nonZeroIdx) {
                nums[zeroIdx] = nums[nonZeroIdx].also{
                    nums[nonZeroIdx] = nums[zeroIdx]
                }
                zeroIdx += 1   
            }
            nonZeroIdx += 1
            
            while(zeroIdx < nums.size && nums[zeroIdx] != 0) {
                zeroIdx += 1
            } 

            while(nonZeroIdx < nums.size && nums[nonZeroIdx] == 0) {
                nonZeroIdx += 1
            }

            if(zeroIdx >= nums.size || nonZeroIdx >= nums.size) {
                return
            }
        }
    }
}